lightning-invoice/src/payment.rs

```diff
@@ -425,28 +496,94 @@ where
 
 		let payer = self.payer.node_id();
 		let first_hops = self.payer.first_hops();
-		let route = self.router.find_route(
-			&payer, params, &payment_hash, Some(&first_hops.iter().collect::<Vec<_>>()),
-			&self.scorer.lock()
-		).map_err(|e| PaymentError::Routing(e))?;
+		let inflight_htlcs = {
+			// Make an attempt at finding existing payment information from `payment_cache`. If it
+			// does not exist, it probably is a fresh payment and we can just return an empty
+			// HashMap.
+			if let Some(payment_info) = self.payment_cache.lock().unwrap().get(&payment_hash) {
+				let mut total_inflight_map: HashMap<(u64, bool), u64> = HashMap::new();
+
+				for hops in &payment_info.paths {
+					if hops.len() < 1 { break };
+					// total_inflight_map needs to be direction-sensitive when keeping track of the HTLC value
+					// that is held up. However, the `hops` array, which is a path returned by `find_route` in
+					// the router excludes the payer node. In the following lines, the payer's information is
+					// hardcoded with an inflight value of 0 so that we can correctly represent the first hop
+					// in our sliding window of two.
+					let our_node_id: PublicKey = self.payer.node_id();
+					let reversed_hops_with_payer = core::iter::once((0u64, our_node_id)).chain(
+						hops.split_last().unwrap().1.iter().map(|hop| (hop.fee_msat, hop.pubkey))).rev();
+					let mut cumulative_msat = hops.last().unwrap().fee_msat;
+
+					// Taking the reversed vector from above, we zip it with just the reversed hops list to
+					// work "backwards" of the given path, since the last hop's `fee_msat` actually represents
+					// the total amount sent.
+					for (next_hop, prev_hop) in hops.iter().rev().zip(reversed_hops_with_payer) {
+						cumulative_msat += prev_hop.0;
+						total_inflight_map
+							.entry((next_hop.short_channel_id, next_hop.pubkey < prev_hop.1))
+							.and_modify(|used_liquidity_msat| *used_liquidity_msat += cumulative_msat)
+							.or_insert(cumulative_msat);
+					}
+				}
+
+				total_inflight_map
+			} else {
+				HashMap::new()
+			}
+		};
+
+		let route = {
+			let scorer = AccountForInFlightHtlcs { scorer: &mut self.scorer.lock(), inflight_htlcs };
+
+			self.router.find_route(
+				&payer, &params, &payment_hash, Some(&first_hops.iter().collect::<Vec<_>>()),
+				&scorer
```

Comment:
FWIW, instead of having a `scorer` variable, you can inline the creation of `AccountForInFlightHtlcs` to avoid needing the scope.